backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/ZonedDateTimeCodec.java

```diff
@@ -0,0 +1,171 @@
+import com.datastax.driver.core.ParseUtils;
+import com.datastax.driver.core.ProtocolVersion;
+import com.datastax.driver.core.TypeCodec;
+import com.datastax.driver.core.UserType;
+import com.datastax.driver.core.exceptions.InvalidTypeException;
+
+public class ZonedDateTimeCodec extends TypeCodec.AbstractUDTCodec<ZonedDateTime> {
+
+    private static final java.time.format.DateTimeFormatter FORMATTER = new java.time.format.DateTimeFormatterBuilder()
+    .parseCaseSensitive()
+    .parseStrict()
+    .append(java.time.format.DateTimeFormatter.ISO_LOCAL_DATE)
+    .optionalStart()
+    .appendLiteral('T')
+    .appendValue(java.time.temporal.ChronoField.HOUR_OF_DAY, 2)
+    .appendLiteral(':')
+    .appendValue(java.time.temporal.ChronoField.MINUTE_OF_HOUR, 2)
+    .optionalEnd()
+    .optionalStart()
+    .appendLiteral(':')
+    .appendValue(java.time.temporal.ChronoField.SECOND_OF_MINUTE, 2)
+    .optionalEnd()
+    .optionalStart()
+    .appendFraction(java.time.temporal.ChronoField.NANO_OF_SECOND, 0, 9, true)
+    .optionalEnd()
+    .optionalStart()
+    .appendZoneOrOffsetId()
+    .optionalEnd()
+    .toFormatter()
+    .withZone(java.time.ZoneOffset.UTC);
+
+    public ZonedDateTimeCodec(UserType definition) {
+        super(definition, java.time.ZonedDateTime.class);
+        checkArgument(definition.getFieldType(CassandraZonedDateTimeModule.DATE).equals(DataType.timestamp()),
+            "Expected UDT with field %s being a timestamp. Got %s",
+            CassandraZonedDateTimeModule.DATE,
+            definition.getFieldType(CassandraZonedDateTimeModule.DATE));
+        checkArgument(definition.getFieldType(CassandraZonedDateTimeModule.TIME_ZONE).equals(DataType.varchar()),
+            "Expected UDT with field %s being a text. Got %s",
+            CassandraZonedDateTimeModule.TIME_ZONE,
+            definition.getFieldType(CassandraZonedDateTimeModule.TIME_ZONE));
+    }
+
+    @Override
+    protected ZonedDateTime newInstance() {
+        return null;
+    }
+
+    @Override
+    protected ByteBuffer serializeField(ZonedDateTime zonedDateTime, String fieldName, ProtocolVersion protocolVersion) {
+        switch (fieldName) {
+            case CassandraZonedDateTimeModule.DATE:
+                return timestamp().serialize(new Date(zonedDateTime.toInstant().toEpochMilli()), protocolVersion);
+            case CassandraZonedDateTimeModule.TIME_ZONE:
+                return varchar().serialize(zonedDateTime.getZone().getId(), protocolVersion);
+            default:
+                throw new InvalidTypeException(fieldName + " is not known as a UDT field used for ZonedDateTime");
+        }
+    }
+
+    @Override
+    protected ZonedDateTime deserializeAndSetField(ByteBuffer byteBuffer, ZonedDateTime zonedDateTime, String fieldName, ProtocolVersion protocolVersion) {
+        switch (fieldName) {
+            case CassandraZonedDateTimeModule.DATE:
+                return ZonedDateTime.ofInstant(
```

Comment:
I will. "protected" method so called from the super class. I have guaranties on the UDT type about field ordering and the caller iterate over fields in order. Want me to add a Preconditions to ensure Date is called first ?

(Apparently even datastax has API calls that looks uggly compared to ours...) 

https://github.com/datastax/java-driver/blob/3.0/driver-extras/src/main/java/com/datastax/driver/extras/codecs/jdk8/ZonedDateTimeCodec.java # deserializeAndSetField

```
        if (index == 0) {
            long millis = bigint().deserializeNoBoxing(input, protocolVersion);
            return Instant.ofEpochMilli(millis).atZone(ZoneOffset.UTC);
        }
        if (index == 1) {
            String zoneId = varchar().deserialize(input, protocolVersion);
            return target.withZoneSameInstant(ZoneId.of(zoneId));
        }
        throw new IndexOutOfBoundsException("Tuple index out of bounds. " + index);
```

Ouch ;-)